vllm/model_executor/models/llava.py

```diff
@@ -0,0 +1,245 @@
+            projector_hidden_act=config.projector_hidden_act)
+
+        self.linear_method = linear_method
+        self.language_model = LlamaModel(hf_text_config,
+                                         linear_method,
+                                         lora_config=lora_config)
+        self.unpadded_vocab_size = hf_text_config.vocab_size
+        self.lm_head = ParallelLMHead(
+            self.unpadded_vocab_size,
+            hf_text_config.hidden_size,
+            org_num_embeddings=self.language_model.org_vocab_size)
+        self.sampler = Sampler(self.unpadded_vocab_size,
+                               self.language_model.org_vocab_size)
+
+    def forward(self,
+                input_ids: torch.Tensor,
+                positions: torch.Tensor,
+                kv_caches: List[KVCache],
+                input_metadata: InputMetadata,
+                image_input: Optional[torch.Tensor] = None) -> SamplerOutput:
+        """Run forward pass for Llava.
+
+        One key thing to understand is the `input_ids` already accounts for the
+        positions of the to-be-inserted image embeddings.
+        Concretely, consider a text prompt:
+        "<image>\nUSER: What's the content of the image?\nASSISTANT:".
+        Tokenizer outputs:
+        [1, 32000, 29871, 13, 11889, 29901, 1724, 29915, 29879, 278,
+        2793, 310, 278, 1967, 29973, 13, 22933, 9047, 13566, 29901].
+        The to-be-inserted image has a size of 576 (24 * 24) along the context
+        length dimension.
+        `input_ids` is thus [1, 32000, ..., 32000, 29871, 13, 11889, 29901,
+        1724, 29915, 29879, 278, 2793, 310, 278, 1967, 29973, 13, 22933,
+        9047, 13566, 29901].
+        There will be 576 `32000` in the `input_ids`.
+        (32000 is the token id for `<image>`.)
+
+        This way, the `positions` and `input_metadata` are consistent
+        with the `input_ids`.
+
+        The design is chosen so that all data massaging happens inside
+        Llava model.
+
+        Args:
+            input_ids: Flattened (concatenated) input_ids corresponding to a
+                batch.
+            image_input: A batch of image inputs (either of type PIXEL_VALUES
+                or IMAGE_FEATURES).
+                For PIXEL_VALUES, expecting [1, 3, 336, 336].
+                For IMAGE_FEATURES, expecting [1, 576, 1024].
+        """
+        if image_input is not None:
+            # This should be ensured by ray-llm.
+            # In case vllm entrypoint is used, this is a user facing error.
+            if list(image_input.shape[1:]) != list(
+                    self.vision_language_config.image_input_shape[1:]):
+                raise ValueError(
+                    f"The expected image tensor shape is batch dimension "
+                    f"plus "
+                    f"{self.vision_language_config.image_input_shape[1:]}."
+                    f" You supplied {image_input.shape[1:]}. "
```

Comment:
Here, you should print the full shape, i.e. `image_input.shape`, otherwise it will be confusing :)